File: src/main/java/io/github/saneea/dvh/FeatureResources.java
package io.github.saneea.dvh;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintStream;
import java.io.Reader;
import java.io.StringWriter;
import java.io.Writer;
import java.nio.charset.Charset;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.stream.Stream;

import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.Options;

import io.github.saneea.dvh.Feature.CLI.CommonOptions;
import io.github.saneea.dvh.Feature.Util.IOConsumer;
import io.github.saneea.dvh.utils.Utils;

public class FeatureResources implements AutoCloseable {

	private final Deque<AutoCloseable> closeables = new ArrayDeque<>();

	private final Feature feature;
	private final String[] args;
	private final FeatureContext context;

	private Options cliOptions;
	private CommandLine commandLine;
	private PrintStream outTextPrintStream;
	private Writer outTextWriter;
	private OutputStream outBinStream;
	private OutputStream errBinStream;
	private Reader inTextReader;
	private String inTextString;
	private InputStream inBinStream;

	public FeatureResources(Feature feature, String[] args, FeatureContext context) {
		this.feature = feature;
		this.args = args;
		this.context = context;
	}

	public Options getCliOptions() {
		if (cliOptions == null) {
			cliOptions = new Options();

			if (feature instanceof Feature.CLI.Options) {
				Stream//
						.of(//
								((Feature.CLI.Options) feature)//
										.getOptions())//
						.forEach(//
								cliOptions::addOption);
			}

			cliOptions.addOption(CommonOptions.HELP_OPTION);

			if (feature instanceof Feature.Out.Text.PrintStream//
					|| feature instanceof Feature.Out.Text.Writer//
					|| feature instanceof Feature.Out.Text.String) {
				cliOptions.addOption(CommonOptions.OUTPUT_ENCODING_OPTION);
			}

			if (feature instanceof Feature.In.Text.Reader//
					|| feature instanceof Feature.In.Text.String) {
				cliOptions.addOption(CommonOptions.INPUT_ENCODING_OPTION);
			}

			if (feature instanceof Feature.In.Bin.Stream//
					|| feature instanceof Feature.Out.Bin.Stream) {
				cliOptions.addOption(CommonOptions.NON_BUFFERED_STREAMS_OPTION);
			}

		}
		return cliOptions;
	}

	public CommandLine getCommandLine() {
		if (commandLine == null) {
			commandLine = Utils.parseCli(args, getCliOptions(), feature, context);
		}
		return commandLine;
	}

	public IOConsumer<String> getOutTextString() {
		return getOutTextWriter()::write;
	}

	public Writer getOutTextWriter() {
		if (outTextWriter == null) {
			outTextWriter = new OutputStreamWriter(//
					getOutBinStream(), //
					getEncoding(CommonOptions.OUTPUT_ENCODING));
			closeables.add(outTextWriter);
		}
		return outTextWriter;
	}

	public PrintStream getOutTextPrintStream() {
		if (outTextPrintStream == null) {
			outTextPrintStream = new PrintStream(//
					getOutBinStream(), //
					false, //
					getEncoding(CommonOptions.OUTPUT_ENCODING));

			closeables.add(outTextPrintStream);
		}
		return outTextPrintStream;
	}

	private Charset getEncoding(String encodingOptionName) {
		String encodingName = getCommandLine().getOptionValue(encodingOptionName);
		return encodingName != null//
				? Charset.forName(encodingName)//
				: Charset.defaultCharset();
	}

	public OutputStream getOutBinStream() {
		if (outBinStream == null) {
			outBinStream = new FileOutputStream(FileDescriptor.out);

			if (useBufferedStreams()) {
				outBinStream = new BufferedOutputStream(outBinStream);
			}

			closeables.add(outBinStream);
		}
		return outBinStream;
	}

	public OutputStream getErrBinStream() {
		if (errBinStream == null) {
			errBinStream = new FileOutputStream(FileDescriptor.err);
			closeables.add(errBinStream);
		}
		return errBinStream;
	}

	public String getInTextString() throws IOException {
		if (inTextString == null) {
			inTextString = readAll(getInTextReader());
		}
		return inTextString;
	}

	private static String readAll(Reader reader) throws IOException {
		StringWriter buf = new StringWriter();
		reader.transferTo(buf);
		return buf.toString();
	}

	public Reader getInTextReader() throws IOException {
		if (inTextReader == null) {
			inTextReader = Utils.skipBom(//
					new InputStreamReader(//
							getInBinStream(), //
							getEncoding(CommonOptions.INPUT_ENCODING)));
			closeables.add(inTextReader);
		}
		return inTextReader;
	}

	public InputStream getInBinStream() {
		if (inBinStream == null) {
			inBinStream = new FileInputStream(FileDescriptor.in);

			if (useBufferedStreams()) {
				inBinStream = new BufferedInputStream(inBinStream);
			}

			closeables.add(inBinStream);
		}
		return inBinStream;
	}

	private boolean useBufferedStreams() {
		return !getCommandLine().hasOption(CommonOptions.NON_BUFFERED_STREAMS);
	}

	@Override
	public void close() throws Exception {
		Exception onCloseException = null;

		while (!closeables.isEmpty()) {
			AutoCloseable closeable = closeables.pollLast();
			try {
				closeable.close();
			} catch (Exception e) {
				if (onCloseException != null) {
					e.addSuppressed(onCloseException);
				}
				onCloseException = e;
			}
		}

		if (onCloseException != null) {
			throw onCloseException;
		}
	}

}

File: src/main/kotlin/io/github/saneea/dvh/FeatureResources.kt
package io.github.saneea.dvh

import io.github.saneea.dvh.Feature.CLI
import io.github.saneea.dvh.Feature.CLI.CommonOptions
import io.github.saneea.dvh.utils.ByteSequenceRecognizer
import io.github.saneea.dvh.utils.Utils
import io.github.saneea.dvh.utils.databuffer.WriteOnCloseOutputStream
import io.github.saneea.dvh.utils.encodingRecognizer
import org.apache.commons.cli.CommandLine
import org.apache.commons.cli.Options
import java.io.*
import java.nio.charset.Charset
import java.util.*

class FeatureResources(
    private val feature: Feature,
    private val context: FeatureContext
) : AutoCloseable {

    private val closeables: Deque<AutoCloseable> = ArrayDeque()
    private val exceptionListeners: MutableSet<(Exception) -> Unit> = LinkedHashSet()

    val commandLine: CommandLine by lazy {
        Utils.parseCli(cliOptions, feature, context)
    }

    val outTextPrintStream: PrintStream by lazy {
        PrintStream(outBinStream, false, getOutputEncoding())
            .also(closeables::add)
    }

    val outTextWriter: Writer by lazy {
        OutputStreamWriter(outBinStream, getOutputEncoding())
            .also(closeables::add)
    }

    val outBinStream: OutputStream by lazy {
        createInternalOutBinStream()
            .also(closeables::add)
    }

    private fun createInternalOutBinStream(): OutputStream {
        val outputFilePath = commandLine.getOptionValue(CommonOptions.OUTPUT_FILE)
        return if (outputFilePath != null)
            WriteOnCloseOutputStream({ FileOutputStream(outputFilePath) })
                .also { exceptionListeners.add { _ -> it.cancel() } }
        else
            FileOutputStream(FileDescriptor.out).let {
                if (useBufferedStreams())
                    BufferedOutputStream(it)
                else
                    it
            }
    }

    val errBinStream: OutputStream by lazy {
        FileOutputStream(FileDescriptor.err)
            .also(closeables::add)
    }

    val inTextReader: Reader by lazy {
        Utils.skipBom(InputStreamReader(inBinStream, inputEncoding))
            .also(closeables::add)
    }

    val inTextString: String by lazy {
        inTextReader.readText()
    }

    private val inputEncoding: Charset by lazy {
        commandLine.getOptionValue(CommonOptions.INPUT_ENCODING)
            ?.let(Charset::forName)
            ?: encodingRecognizer.result
            ?: Charset.defaultCharset()
    }

    private val cliOptions: Options by lazy {
        Options().also {

            if (feature is CLI.Options) {
                feature.options.forEach(it::addOption)
            }

            it.addOption(CommonOptions.HELP_OPTION)

            when (feature) {
                is Feature.Out.Text.PrintStream,
                is Feature.Out.Text.Writer,
                is Feature.Out.Text.String ->
                    it.addOption(CommonOptions.OUTPUT_ENCODING_OPTION)
            }

            when (feature) {
                is Feature.In.Text.Reader,
                is Feature.In.Text.String ->
                    it.addOption(CommonOptions.INPUT_ENCODING_OPTION)
            }

            when (feature) {
                is Feature.In.Bin.Stream,
                is Feature.In.Text.Reader,
                is Feature.Out.Bin.Stream,
                is Feature.Out.Text.PrintStream,
                is Feature.Out.Text.Writer ->
                    it.addOption(CommonOptions.NON_BUFFERED_STREAMS_OPTION)
            }

            when (feature) {
                is Feature.In.Bin.Stream,
                is Feature.In.Text.Reader,
                is Feature.In.Text.String ->
                    it.addOption(CommonOptions.INPUT_FILE_OPTION)
            }

            when (feature) {
                is Feature.Out.Bin.Stream,
                is Feature.Out.Text.Writer,
                is Feature.Out.Text.PrintStream,
                is Feature.Out.Text.String ->
                    it.addOption(CommonOptions.OUTPUT_FILE_OPTION)
            }
        }
    }

    val outTextString: StringConsumer by lazy {
        { str: String -> outTextWriter.write(str) }
    }

    private fun getOutputEncoding() = commandLine
        .getOptionValue(CommonOptions.OUTPUT_ENCODING)
        ?.let(Charset::forName)
        ?: Charset.defaultCharset()

    val inBinStream: InputStream by lazy {
        encodingRecognizer.stream
    }

    private val encodingRecognizer: ByteSequenceRecognizer<Charset> by lazy {
        encodingRecognizer(createInternalInBinStream())
            .also(closeables::add)
    }

    private fun createInternalInBinStream(): InputStream {
        val inputFilePath = commandLine.getOptionValue(CommonOptions.INPUT_FILE)
        return if (inputFilePath != null)
            FileInputStream(inputFilePath)
        else
            FileInputStream(FileDescriptor.`in`).let {
                if (useBufferedStreams())
                    BufferedInputStream(it)
                else
                    it
            }
    }

    private fun useBufferedStreams() = !commandLine.hasOption(CommonOptions.NON_BUFFERED_STREAMS)

    override fun close() {
        var onCloseException: Exception? = null
        while (!closeables.isEmpty()) {
            val closeable = closeables.pollLast()
            try {
                closeable.close()
            } catch (e: Exception) {
                if (onCloseException != null) {
                    e.addSuppressed(onCloseException)
                }
                onCloseException = e
            }
        }
        if (onCloseException != null) {
            throw onCloseException
        }
    }

    fun onException(e: Exception) = exceptionListeners.forEach { it.invoke(e) }
}
